### ISSUE TITLE
[in_app_purchase] Unable to get recent Android purchase details

### ISSUE THREAD

**Issue by lukeirvin** [flutter/flutter#110388]
Flutter Version -- 2.10.5
in_app_purchase: ^3.0.7

I have also tested with:
```
dependency_overrides:
  in_app_purchase_android: 0.2.3+1
```

_Everything below is currently being tested in Android Sandbox Environment._

The issue I am having is when I make a successful subscription purchase on Android it works great. If I am presented with my paywall again, neither Restore nor Purchase retrieves the `purchaseDetail` data on Android. This does work on iOS though.

I have also tried:

```
_checkAndroidPastPurchases() async {
    final InAppPurchaseAndroidPlatformAddition androidAddition = _inAppPurchase
        .getPlatformAddition<InAppPurchaseAndroidPlatformAddition>();
    final QueryPurchaseDetailsResponse oldpurchases =
        await androidAddition.queryPastPurchases();
   
    for (final oldP in oldpurchases.pastPurchases) {
      // pastPurchases is empty
    }
}

```

But I always get an empty response.

So because of this, I can not restore my purchase and if I try to go through the buy process again I see the prompt that states my device already has an active subscription.

I am also getting email notifications when my sandbox subscription renews.

Is there specific logic I need to use to get the known subscription purchase details on my Android device?

flutter doctor:

```
[flutter] flutter doctor -v
[✓] Flutter (Channel stable, 2.10.5, on macOS 12.5 21G72 darwin-arm, locale en-US)
    • Flutter version 2.10.5 at /Users/lukeirvin/Development/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 5464c5bac7 (4 months ago), 2022-04-18 09:55:37 -0700
    • Engine revision 57d3bac3dd
    • Dart version 2.16.2
    • DevTools version 2.9.2

[✓] Android toolchain - develop for Android devices (Android SDK version 32.1.0-rc1)
    • Android SDK at /Users/lukeirvin/Library/Android/sdk
    • Platform android-32, build-tools 32.1.0-rc1
    • Java binary at: /Applications/Android Studio.app/Contents/jre/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.11+0-b60-7772763)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 13.4.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • CocoaPods version 1.11.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2021.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.11+0-b60-7772763)

[✓] VS Code (version 1.70.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.46.0

[✓] Connected device (2 available)
    • SM G998U1 (mobile) • R5CR8041PQW • android-arm64  • Android 12 (API 31)
    • Chrome (web)       • chrome      • web-javascript • Google Chrome 104.0.5112.101

[✓] HTTP Host Availability
    • All required HTTP hosts are available

• No issues found!
exit code 0
```

Update on 8/27/22:
Seems I am always stuck on this line now:

`Unhandled Exception: type 'PurchaseDetails' is not a subtype of type 'GooglePlayPurchaseDetails' in type cast`

My understanding is that PurchaseDetails should know how to set this based on the platform???

So could this be a bug that needs to be fixed?

Or

Do I need to do anything Android-specific when I make my purchase?

**Comment by danagbemava-nc** [comment]
Hi @lukeirvin, can you upgrade to the latest version of `in_app_purchase_android` to see if you still experience the issue querying past purchases? 

According to the [changelog](https://pub.dev/packages/in_app_purchase_android/changelog#0232) this should be fixed in `in_app_purchase_android: 0.2.3+2`

**Comment by lukeirvin** [comment]
> Hi @lukeirvin, can you upgrade to the latest version of `in_app_purchase_android` to see if you still experience the issue querying past purchases?
> 
> According to the [changelog](https://pub.dev/packages/in_app_purchase_android/changelog#0232) this should be fixed in `in_app_purchase_android: 0.2.3+2`

Hi @danagbemava-nc.

I believe my issue with Android is working, but iOS isn't now. 

Mainly I either get stuck with a PaymentQueue full of past restored items or nothing is returned at all & I can't detect if it is a canceled status. 

Or, do you know if there are any differences between detecting a cancel status for iOS in Sandbox vs. Production?

**Comment by danagbemava-nc** [comment]
Hi @lukeirvin, kindly file a new issue for iOS so that it can be tracked separately. Please provide all the relevant information when you file the issue. 

> Or, do you know if there are any differences between detecting a cancel status for iOS in Sandbox vs. Production?

As far as I know, there shouldn't be any differences in detecting the cancel status on sandbox vs production 

> I believe my issue with Android is working,

Given this, I believe it is safe to close this issue. 

If you do run into this issue again, kindly leave a comment below and I will reopen this. 

Thank you
